app/assets/javascripts/views/hierarchical_response_condition_checker.js.coffee

```diff
@@ -0,0 +1,133 @@
+
+  initialize: (options) ->
+    @manager = options.manager
+    @condition = options.condition
+
+    @rqElement = @manager.qingElement(@condition.refQingId)
+    @rqType = @rqElement.data('qtype-name')
+    @result = true
+
+    # These handlers must be set dynamically based on rqElement.
+    @rqElement.find('div.control').find('input, select, textarea').on('change', => @checkAndTell())
+    @rqElement.find("div.control input[type=text]").on('keyup', => @checkAndTell())
+    @rqElement.find("div.control input[type=number]").on('keyup', => @checkAndTell())
+    if @rqType == 'long_text' && !@manager.readOnly
+      @ckeditorInstance().on('change', => @checkAndTell())
+
+    # Set result to initial value on page load. Don't tell the manager because the manager calls refresh just
+    # once on page load (if we 'tell' when each checker initializes, the manager would evaluate many times)
+    @check()
+
+  checkAndTell: ->
+    @check()
+    @manager.refresh()
+
+  evaluate: ->
+    @result
+
+  # Checks the result of the condition and sets the result value.
+  check: ->
+    unless @rqElement.is(':visible')
+      @result = false
+      return
+
+    actual = @actual()
+    expected = @expected()
+
+    # For select_one questions, the actual will be an array of selected option_node_ids.
+    # We should return true if the expected option_node_id is anywhere in that list.
+    # So `eq` just becomes `inc` and `neq` becomes `ninc`.
+    # We could label it this way in the condition form but it seems that would be confusing.
+    op = @condition.op
+    if @rqType == 'select_one'
+      op = 'inc' if op == 'eq'
+      op = 'ninc' if op == 'neq'
+
+    @result = switch op
+      when 'eq' then @testEquality(actual, expected)
+      when 'lt' then actual < expected
+      when 'gt' then actual > expected
+      when 'leq' then actual <= expected
+      when 'geq' then actual >= expected
+      when 'neq' then @testInequality(actual, expected)
+      when 'inc' then actual.indexOf(expected) != -1
+      when 'ninc' then actual.indexOf(expected) == -1
+      else false
+
+  # Uses a special array comparison method if appropriate.
+  testEquality: (a, b) ->
+    if $.isArray(a) && $.isArray(b)
+      a.equalsArray(b)
+    else
+      a == b
+
+  # For inequality conditions, ignore nulls for relevance test
+  # unless null is the value being checked by the condition
+  # improves UX for appearing/disappearing questions.
+  testInequality: (refValue, expected) ->
+    if $.isArray(refValue) && $.isArray(expected)
+      if refValue.equalsArray([]) && !expected.equalsArray([])
+        false
+      else
+        !refValue.equalsArray(expected)
+    else
+      # If you aren't checking for "not null"
+      if refValue == null && expected != null
+        false
+      else
+        refValue != expected
+
+  # Gets the actual answer for the referred question.
+  actual: ->
+    # If readonly, use the data-val attrib
+    if @manager.readOnly
```

Comment:
We don't need readOnly any more, or any of the .ro-val stuff. Can you strip all that out? Conditions are now only useable when you're creating/editing a response.